### ISSUE TITLE
Add debug info feature to CE native image generator

### ISSUE THREAD

**Issue by adinn** [oracle/graal#1917]
Red hat would like to work with the GraalVM project to add debug info to the native image generator for use by command line debuggers and IDEs. The generated debug info should include:

 - symbols for primitive, object and interface types
 - symbols for methods
 - size and numeric type info for Java primitive data
 - size and layout info for Java object data
 - address ranges for method code
 - file and line number info for specific code addresses within methods
 - method frame layouts at end of method prologue and at specific addresses within calls
 - parameter and local var type and location info (register/stack/memory) at end of method prologue and at specific addresses within calls
 - address and type for Java primitive and object values located in the heap region

We believe it should be possible to use this info to support the following iniital debugging functionality:

 - break points configured by file and line or by method name
 - single stepping by line including both into and over function calls
 - stack backtraces (not necessarily including frames detailing inlined code)
 - structured display of primitive values and objects
 - including display of data accessed through object networks via path expressions

Debug info should cover both app code and JDK Runtime/SubstrateVM code and types.

The current plan is to implement a working version for ELF/Linux using DWARF info. This will be verified using gdb as the target debugger. This will proof the idea and also identify a model for plugging in the capability as a Native Image Generator feature. A follow up task will attempt to provide equivalent debug info to support denbugging for Windows/PECOFF and MacOS/MACHO images.

**Comment by adinn** [comment]
I'd be grateful if anyone, particularly Oracle staff, would care to comment on the proposed feature list and/or associated debug info that I envisage need to be injected into native images. I have made considerable progress with a prototype implementation of some of these features which I hope to make available soon for early review. It would be very helpful to know what others consider to be priorities for that implementation.

n.b. I am afraid I have no idea what features -- or, indeed, debug info --  the product code provides (I am deliberately not looking at the product behaviour or image object structure for reasons that I assume are obvious).

**Comment by dougxc** [comment]
The proposed feature set and the debugging capabilities it provides look like a very good start. I spoken a bit with @olpaw on this and some design considerations fell out of our conversation:
* Native image debug info should be modeled as independently as possible from a specific debug info format such as DWARF. Only in the backend should this model be converted to DWARF/PECOFF etc.
* You should use `org.graalvm.compiler.code.CompilationResult.getSourceMappings()` instead of `org.graalvm.compiler.code.CompilationResult.getInfopoints()` in terms of getting source line info.
* In terms of debugging experience, it might be worth considering debugger extensions (e.g. using gdb's python API) tha do things such as:
  * Simplify setting breakpoints based on Java names. For example: `> ni-bp java.lang.Class.isAnonymousClass(java.lang.Object)boolean`
  * pretty print values based on dynamic info. For example, DWARF may state that the static type of a variable is `java.lang.Object` but it may actually be `java.util.ArrayList` and reading through the object's header, better pretty printing is possible.
* You should also consider what's possible in terms of native debug info for runtime generated code (e.g. Truffle compilation in a native image runtime).

**Comment by olpaw** [comment]
* `CompilationResult.getSourceMappings()` should give you what you need for `.debug_info` and `.debug_line`
* `org.graalvm.compiler.code.CompilationResult#getMarks()` can be used to build `.debug_frame` (use the `MARK_PROLOGUE/EPILOGUE` strings in SubstrateBackend derived classes as keys)
* To generate representations for types (`DW_AT_type` entries) you can make use of `com.oracle.svm.hosted.meta.HostedUniverse#getTypes()`

**Comment by christianwimmer** [comment]
Windows will be a supported platform for Native Image very soon, so it would be good to have Windows support from the beginning and not as a "follow up task".

If you plan the rely on any external library, please let us know as early as possible because all external dependencies have to go through an extensive approval process.

**Comment by adinn** [comment]
Thanks very much for the extremely helpful feedback.

I had already identified CompilationResult.getSourceMappings as the route to derive line info and HostedUniverse.getTypes() to derive type info. My current prototype uses the former to generate enough DWARF content to support line and method name based breakpoints and line-by-line stepping (i.e. so far I have abbrev, info, range and line sections). The info section does not yet locate methods records as members of an associated class as I have not added type info.

I have started looking into generating details of classes and I believe HostedUniverse.getTypes() will provide most of what is needed to generate the extra DWARF records needed to support most of the remaining functions.

Before moving on to cater for type info I need to complete derivation of debug info from the the CompilationResult suite with one last feature -- adding frame base fields to the DWARF info records that describe methods. That should be enough to allow stack backtraces to work (and also, eventually, allow specification of stack locations for locals relative to the frame base). I had already spotted CompilationResult.totalFrameSize() but had not yet seen CompilationResult.getMarks(). Clearly, that is is all I need in order to allow the debugger to identify the frame base. So, I also hope to have stack backtraces working in gdb fairly soon. I think it might be wise to post a PR for review once that is working.

As regards future work, I would very much like to have Windows support as well as DWARF for Linux in whatever gets committed. One of my colleagues is looking into implementing this. He is currently using Visual Studio and tailoring the PECOFF debug info that he generates to what it expects. Does the project assume a specific tool suite and version?

Regarding the front end/back end split, I understand this constraint. Indeed working around it was one of the trickiest things I needed to puzzle out in order to get started. I have located all the logic for generating DWARF debug within the same package as ElfObjectFile. Indeed, at the moment is is implemented using a single class called DwarfSections within which I have embedded static member classes that implement each relevant DWARF section as a suitable subclass of ElfSection. Class DwarfSections constructs a shared model of the underlying type/code/heap which the individual sections can reference in order to size and build their own content. Similarly, my colleague has located all the logic for generating VS debug info into PECoffObjectFile.

Obviously, making this work has required providing an abstraction layer on ObjectFile which allows the details of the underlying method, type and heap layouts to be presented to the ObjectFile implementation without the latter knowing anything about the underlying Substrate/Graal types. I have managed to implement such a split by adding a suite of interfaces to ObjectFile which allow Type, Code and Heap info to be queried in a generic way. These interfaces are implemented by classes defined in NativeBootImage in much the same way as it provides/overrides NativeTextSectionImpl. My DwarfSections class uses the interface API to interrogate the supplied implementations ad build its own internal model of the types, code and heap.

I am sure the way I have achieved this  i) is less clean than is desirable and ii) does not conform to the model Graal employs for implementing add-on features. However, it does ensure that no change are needed to type visibility for the current various different Graal source modules/packages. No doubt when my prototype is ready for review you will have ample opportunities for feedback to help improve this decoupling :-)

Another interesting point of concern is the AArch64/x86_64 split. Provision of frame info is the first occurrence up to now in my prototype where the underlying hardware architecture influences (DWARF) debug info generation. On Linux/x86 I need to encode the frame base as an offset relative to RSP/RBP and on Linux/AArch64 it has to be relative to RSP/RFP. That requires generating different location codes because registers need to be referred to by index number (r7/r8 vs r31/r29). I believe I can finesse this and other such differences in generated content by delegating such occurrences to an arch-specific generator subclass internal to the DWARF generator suite, instantiating it modulo the ElfObjectFile arch setting.

Regarding library support I was not intending to use any external libraries. So far, I have used pure Java to write all DWARF content. Of course, the whole of DWARF is huge. However, implementing the subset needed to support debugging for native images ought not to be intractable. I don't think my colleague has any plans to use lbs for the Windows code but I will check that with him and get back to you.

I have looked into gdb's Python API in the past and actually managed to do so without resorting to fleeing the country and joining the Foreign Legion. It would indeed be nice, say, to specify method names using Java syntax or print an Object as an ArrayList where the relevant type inclusion relation held. I thought I could maybe leave this sort of fancy command processing to someone who actually knows and likes Python (hmm, ... Doug ... :-).

The possibility of generating debug info for gdb on the fly to handle dynamic compilation does sound very appealing. I am not sure whether the current gdb API to hand over DWARF info is actually ready for prime-time. I had a look into it a couple of years ago when trying to extend gdb to do stack backtraces for Hotspot and could not find an easy way to use it (despite talking through the problems with gdb devs). In the end I found it much easier to use the Python API to extend gdb so that it knew how to traverse the code heap and other Hotspot internal data structures in order to identify details of compiled methods. I will have a chat with some of the gdb devs to find out if anything has improved in the latest implementation.

**Comment by olpaw** [comment]
>  I am not sure whether the current gdb API to hand over DWARF info is actually ready for prime-time.

Yes, this is somewhat wacky. You have to use the interfaces defined in
https://sourceware.org/gdb/current/onlinedocs/gdb/Declarations.html#Declarations
to communicate runtime-compiled code created by the inferior to GDB. You can use it to point GDB to a dwarf section that you create in memory that corresponds to the debuginfo for the newly created code.

**Comment by dougxc** [comment]
> I thought I could maybe leave this sort of fancy command processing to someone who actually knows and likes Python (hmm, ... Doug ... :-)

There could well be some existing work in this area that can be contributed once your initial implementation is in place ;-)

**Comment by stooke** [comment]
I am a colleague of Andrew Dinn's at Red Hat, and I am looking at adding Windows/PECOFF debug information support to the native image generator. 

My target is the Visual Studio debugger; I would like to have the IDE treat Graal programs as first class citizens, with full breakpoint and single stepping support.  This means I need to emit CodeView 4 debug information in the object file.  Microsoft has stated they will be ignoring pure COFF debug information at some point in the future, but that seems tricky, without a new executable format.

The CodeView 4 format is poorly documented, mostly by source code.  I am writing my own implementation to support this, and have made fair progress, although I have yet to successfully set a breakpoint in a java function (an issue with relocations).

I will be following Andrew's PR with interest, and have already made modifications to mine based on the feedback he has received.  Eventually I hope to submit a similar PR for the Windows/PECOFF side of this effort.

**Comment by bobvandette** [comment]
I assume you have been following the llvm work to add support for Visual Studio debugging.

https://llvm.org/devmtg/2016-11/Slides/Kleckner-CodeViewInLLVM.pdf

The LLVM project has done some research and with the help of Microsoft has documented the PDB file format.  

https://llvm.org/docs/PDB/index.html
https://github.com/Microsoft/microsoft-pdb

**Comment by stooke** [comment]
> I assume you have been following the llvm work to add support for Visual Studio debugging.
> 
Yes, those are actually my two primary sources for my work. Currently I emit file and line number records, function entry definitions, and relocation records (via the Graal framework) for those records.  While there are quite a few tools available to parse PECOFF files, the actual internals of Visual Studio are opaque.

I can actually set a breakpoint in a member function (aliased to a global name for the moment), but the source code doesn't come up and single stepping doesn't yet work.  To get to that step without those sources would have been impossible - for me, at least.

**Comment by a74nh** [comment]
(Arm employee, been working on GDB for a few years, now starting to look at Graal)

GDB used to support debugging Java programs compiled by GCJ. However all that support has been removed:

https://github.com/bminor/binutils-gdb/commit/9c37b5aed98e5996a9777a366bfcc371c0e1a92d
```
Remove Java support

This patch removes the Java support from gdb.  gcj has not seen much
development or use for years now, and was recently removed from GCC.

This patch changes gdb to follow; in the unlikely event that there are
still users using gcj, they can continue to use an older gdb to debug.
Or, they can debug in C++ mode.

Built and regtested on x86-64 Fedora 24.

2016-10-06  Tom Tromey  <tom@tromey.com>
```

The last release before this patch was applied was 7.12.1 - released 2017-01. Given most distros have older GDB versions than this, you'll probably be fine... for now.

Almost all the code removed by the patch is just special formatting for Java. So, with the latest GDB, just pretending you are C++ _should_ get you quite far down the road.

If you wanted to get support back into GDB Head....

As expected, the patch doesn't cleanly revert. Mostly in dwarf2read.c and symtab.c, which are fairly fundamental files. So, some work required to resurrect it.

Would the GDB community accept patches to add support to debug Graal native images? The sticking point I think is that the above patch is a lot of code, that they'd then need to support going forward. They'd need convincing Graal native images are here to stay.

**Comment by adinn** [comment]
Hi Alan,

Thanks very much for your input. I fully understand that the gdb community may want to see some more traction before committing to what might be a lot of work on better gdb support for Graal native images. As you say, we should still be able to get a lot of debug functionality to work simply by faking Java as a C++ variant, pirating it's debug info layouts. Let's see how that goes for now.

**Comment by adinn** [comment]
Talking through the initial prototype implementation with @olpaw we have arrived a further requirement regarding making source files available to the debugger. The basic idea is to assemble the necessary source base at image generation time.

Source files for JDK runtime, GraalVM and application classes should be located wherever possible at generation time. Copies of these source files should be assembled under a directory named `sources` residing in the same directory as the generated image.

When source files can be so located then file paths inserted into the debug info can be constructed to identify the associated source file relative to this sources directory.

  - JDK class file sources should be locatable in the src.zip file located in the current java image's lib dir.

  - GraalVM sources should be locatable via the class path:
  jar entries like path/to/xxx.jar should have an associated source jar path/to/xxx.src.zip; 
  dir entries like path/to/dir should have associated source dirs path/to/dir/src and path/to/dir/src_gen

  - Application jars or dirs in the classpath will need to be resolved to source jars or dirs using yet to be defined strategies. As one possibility, jars downloaded via maven ought to be able to be resolved to source jars adjacent to the deployed jar using the standard maven naming convention for source jars relative to their original class file jar.

In all cases where a source file is located it needs to be copied to an appropriate subdir of the the sources dir. The relative path to the source file should include a path prefix determined according to where the source file originated from and should also include a tail derived from the package components. This relative path can be passed to the debug info generator.

In cases where a source file cannot be located the file path should simply be derived from the package components.

n.b. the base file name (excluding the path) for a class's source file will normally be derived from whatever source file name is present in the class file. Where that is not available it can be constructed from the class name.

**Comment by stooke** [comment]
While doing the Windows Codeview implementation, I ran across the same source code issue.  Hidden in the bowels of Visual Studio, there are configuration files that allow the user to manually specify source file search paths, but I see no obvious way to modify these binary files programatically (they're OLE2 files, so it's not hard to write a utility).

It seems to me that providing a Visual Studio Code plugin might actually provide the best cross-platform UX.  I know Graal provides some VSCode plugins for Truffle, for example.  I haven't yet tried the VSCode C++ plugin to see how well native debug can be implemented.

**Comment by adinn** [comment]
The review process has clarified a few more elements in the definition of this feature which I am writing up here to record them and allow discusssion by any interested parties.

1) Two command line options have been identified:

    -H:GenerateDebugInfo=<N>
    -H:DebugInfoSourceSearchPath=<path>

The first option takes an integer values and enables debug info generation if that value is positive (tthe default value is 0). Different positive values are not yet distinguished.

The second option allows root locations for sources to be passed to the source file locator in order for it to populate the sources cache. The supplied path can identify a directory, a sources jar archive (which must have a .jar extension) or a sources zip archive (which must have a .zip extension). This option can appear multiple times to notify multiple source roots. Also, a value provided to this option can identify multiple roots by separating them with a comma.

2) The following rules have been adopted for derivation of application source roots from classpath entries.

If a classpath entry dentifies a jar (e.g. /path/to/foo.jar) or zip file (e.g. /path/to/foo.zip) then the associated source root will be derived by adding qualifier "-sources" to the jar or zip file base name (e.g. path/to/foo-sources.jar or path/to/foo-sources.zip, respectively).

If a classpath entry identifies a directory (e.g. /path/to/foo) then the associated source root will be identified by appending src to that directory (e.g. /path/to/foo/src). However, as a special case, two well known suffixes will be stripped. If the path ends with "/classes" then that suffix will be stripped. If the remaining path ends with "target" then that suffix will also be stripped (so, e.g. /path/to/foo/target/classes, /path/to/foo/classes and /path/to/foo/target will all translate to the same derived source path /path/to/foo/src).

**Comment by stooke** [comment]
I have a preview of my preliminary Windows work available.  Although it will work as-is, it is dependent on Andrew Dinn's work.  I can create PR for this, so it can be discussed and reviewed.  Should I create a separate issue?

**Comment by olpaw** [comment]
>  Should I create a separate issue?

Since github allows more than two parties to interact on one PR you could do with sharing the existing one. If @adinn changes code you depend on too often then I see why you would want to use your own PR. From my experience, two people working on the same branch can be pretty efficient if they are in the same timezone.

**Comment by stooke** [comment]
Although I'm happy to use the same issue, I think it's best if it's a separate PR; Andrew and I are on different continents, and most of my code is completely separate. This also allows his code to be delivered first. Is it okay if I create separate PR?

**Comment by olpaw** [comment]
> Is it okay if I create separate PR?

Sure :-)

**Comment by stooke** [comment]
I have created PR #2245, with a prototype Windows debug implementation.  It is still a work in progress, and is missing types, variables, and complete stack frame records. It will be dependent on @adinn PR #2088.
